libqpdf/QPDF_linearization.cc

```diff
@@ -1058,83 +1054,53 @@ QPDF::calculateLinearizationData(std::map<int, int> const& object_stream_data)
-        std::set<ObjUser>& ous = oiter.second;
-
-        bool in_open_document = false;
-        bool in_first_page = false;
-        int other_pages = 0;
-        int thumbs = 0;
-        int others = 0;
-        bool in_outlines = false;
-        bool is_root = false;
-
-        for (auto const& ou: ous) {
-            switch (ou.ou_type) {
-            case ObjUser::ou_trailer_key:
-                if (ou.key == "/Encrypt") {
-                    in_open_document = true;
-                } else {
-                    ++others;
-                }
-                break;
 
-            case ObjUser::ou_thumb:
-                ++thumbs;
-                break;
+    for (auto& [og, ous]: m->object_to_obj_users) {
+        switch (ous.cbegin()->ou_type) {
+        case ObjUser::ou_other:
+            lc_other.insert(og);
+            break;
 
-            case ObjUser::ou_root_key:
-                if (open_document_keys.count(ou.key) > 0) {
-                    in_open_document = true;
-                } else if (ou.key == "/Outlines") {
-                    in_outlines = true;
-                } else {
-                    ++others;
-                }
-                break;
+        case ObjUser::ou_thumb:
+            if (ous.size() == 1) {
+                lc_thumbnail_private.insert(og);
+            } else {
+                lc_thumbnail_shared.insert(og);
+            }
+            break;
 
-            case ObjUser::ou_page:
-                if (ou.pageno == 0) {
-                    in_first_page = true;
-                } else {
-                    ++other_pages;
-                }
-                break;
+        case ObjUser::ou_open_doc:
+            lc_open_document.insert(og);
+            break;
 
-            case ObjUser::ou_root:
-                is_root = true;
-                break;
+        case ObjUser::ou_outlines:
+            lc_outlines.insert(og);
+            break;
 
-            case ObjUser::ou_bad:
-                stopOnError("INTERNAL ERROR: QPDF::calculateLinearizationData: "
-                            "invalid user type");
-                break;
+        case ObjUser::ou_first_page:
+            if (ous.size() == 1) {
+                lc_first_page_private.insert(og);
+            } else {
+                lc_first_page_shared.insert(og);
             }
-        }
+            break;
 
-        if (is_root) {
-            lc_root.insert(og);
-        } else if (in_outlines) {
-            lc_outlines.insert(og);
-        } else if (in_open_document) {
-            lc_open_document.insert(og);
-        } else if ((in_first_page) && (others == 0) && (other_pages == 0) && (thumbs == 0)) {
-            lc_first_page_private.insert(og);
-        } else if (in_first_page) {
-            lc_first_page_shared.insert(og);
-        } else if ((other_pages == 1) && (others == 0) && (thumbs == 0)) {
-            lc_other_page_private.insert(og);
-        } else if (other_pages > 1) {
```

Comment:
@jberkenbilt 

I think the test above should be ```other_pages > 0```. As it stands, if 
```other_pages == 1 && (others > 0 || thumbs > 0)```
the object ends up in part9 rather than part6.   This doesn't make sense to me and is also inconsistent with how shared objects refered to by the first page are handled.

Reply:
On reflection, I think the problem is as described in #997